### ISSUE TITLE
[webpack-cli] RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by dagouzhi** [callstack/repack#642]
When configuring Webpack with LimitChunkCountPlugin, it reports "Maximum call stack size exceeded."

```
new webpack.optimize.LimitChunkCountPlugin({
  maxChunks: 2
})
@callstack/repack/src/webpack/plugins/OutputPlugin.ts:234:9) at Array.flatMap (<anonymous>)
```
Version 3.7.0 works fine.

**Comment by jbroma** [comment]
Hi @dagouzhi, thanks for reporting the issue.

I'll try to reproduce the error on my own but it does look like the new implementation of OutputPlugin (introduced in v4) didn't take circular references from webpack into account 👍

**Comment by aw-capgemini** [comment]
I recently migrated from Repack v3 to v4. However, when I execute pnpm bundle to create new chunks, I encounter the same error:

RangeError: Maximum call stack size exceeded

@jbroma do we have a fix for this?

**Comment by dagouzhi** [comment]
I've reverted `webpack/plugins/OutputPlugin` back to the v3 version, and it can now be used normally.
```
diff --git a/node_modules/@callstack/repack/dist/webpack/plugins/OutputPlugin.js b/node_modules/@callstack/repack/dist/webpack/plugins/OutputPlugin.js
index 3483493..0161aba 100644
--- a/node_modules/@callstack/repack/dist/webpack/plugins/OutputPlugin.js
+++ b/node_modules/@callstack/repack/dist/webpack/plugins/OutputPlugin.js
@@ -4,34 +4,16 @@ Object.defineProperty(exports, "__esModule", {
   value: true
 });
 exports.OutputPlugin = void 0;
-var _path = _interopRequireDefault(require("path"));
-var _webpack = _interopRequireDefault(require("webpack"));
-var _AssetsCopyProcessor = require("./utils/AssetsCopyProcessor");
-var _AuxiliaryAssetsCopyProcessor = require("./utils/AuxiliaryAssetsCopyProcessor");
-function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
-/**
- * Matching options to check if given {@link DestinationConfig} should be used.
- */
 
-/**
- * Destination config for local chunks.
- */
+var _path = _interopRequireDefault(require("path"));
 
-/**
- * Destination config for remote chunks.
- */
+var _webpack = _interopRequireDefault(require("webpack"));
 
-/**
- * Destination config for chunks.
- */
+var _AssetsCopyProcessor = require("./utils/AssetsCopyProcessor");
 
-/**
- * Destination specification for chunks.
- */
+var _AuxiliaryAssetsCopyProcessor = require("./utils/AuxiliaryAssetsCopyProcessor");
 
-/**
- * {@link OutputPlugin} configuration options.
- */
+function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
 
 /**
  * Plugin for copying generated files (bundle, chunks, assets) from Webpack's built location to the
@@ -48,37 +30,38 @@ class OutputPlugin {
   constructor(config) {
     this.config = config;
     this.config.enabled = this.config.enabled ?? true;
+
     if (!this.config.platform) {
       throw new Error('Missing `platform` option in `OutputPlugin`');
     }
+
     if (!this.config.output) {
       throw new Error('Missing `output` option in `OutputPlugin`');
     }
+
     this.config.extraChunks = this.config.extraChunks ?? [{
       include: /.*/,
       type: 'remote',
       outputPath: _path.default.join(this.config.context, 'build/outputs', this.config.platform, 'remotes')
     }];
   }
-
   /**
    * Apply the plugin.
    *
    * @param compiler Webpack compiler instance.
    */
+
+
   apply(compiler) {
     if (!this.config.enabled) {
       return;
     }
-    const outputPath = compiler.options.output?.path;
-    if (!outputPath) {
-      throw new Error('Cannot infer output path from compilation');
-    }
+
     const logger = compiler.getInfrastructureLogger('RepackOutputPlugin');
-    const extraAssets = (this.config.extraChunks ?? []).map(spec => spec.type === 'remote' ? {
-      ...spec,
+    const extraAssets = (this.config.extraChunks ?? []).map(spec => spec.type === 'remote' ? { ...spec,
       outputPath: !_path.default.isAbsolute(spec.outputPath) ? _path.default.join(this.config.context, spec.outputPath) : spec.outputPath
     } : spec);
+
     const isLocalChunk = chunkId => {
       for (const spec of extraAssets) {
         if (spec.type === 'local') {
@@ -91,85 +74,106 @@ class OutputPlugin {
           }
         }
       }
+
       return false;
     };
-    const getRelatedSourceMap = chunk => {
-      return chunk.auxiliaryFiles?.find(file => /\.map$/.test(file));
-    };
-    const getAllInitialChunks = (chunk, chunks) => {
-      if (!chunk.parents?.length) return [chunk];
-      return chunk.parents.flatMap(parent => {
-        return getAllInitialChunks(chunks.get(parent), chunks);
-      });
-    };
-    compiler.hooks.done.tapPromise('RepackOutputPlugin', async stats => {
-      const compilationStats = stats.toJson({
-        all: false,
-        assets: true,
-        chunks: true,
-        chunkRelations: true,
-        ids: true
-      });
-      const statsChunkMap = new Map(compilationStats.chunks.map(chunk => [chunk.id, chunk]));
-      const entryChunkName = this.config.entryName ?? 'main';
-      const localChunks = [];
-      const remoteChunks = [];
-      const sharedChunks = new Set();
-      const auxiliaryAssets = new Set();
-      const entryChunk = compilationStats.chunks.find(chunk => {
-        return chunk.initial && chunk.names?.includes(entryChunkName);
-      });
-      for (const chunk of compilationStats.chunks) {
-        // Do not process shared chunks right now.
-        if (sharedChunks.has(chunk)) {
-          continue;
-        }
-        getAllInitialChunks(chunk, statsChunkMap).filter(sharedChunk => sharedChunk !== chunk).forEach(sharedChunk => {
-          sharedChunks.add(sharedChunk);
-        });
 
-        // Entry chunk
-        if (entryChunk === chunk) {
-          localChunks.push(chunk);
-        } else if (isLocalChunk(chunk.name ?? chunk.id?.toString())) {
-          localChunks.push(chunk);
-        } else {
-          remoteChunks.push(chunk);
-        }
-      }
+    let entryGroup;
+    let entryChunk;
+    const entryChunkName = this.config.entryName ?? 'main';
+    const localChunks = [];
+    const remoteChunks = [];
+    const auxiliaryAssets = new Set();
+    compiler.hooks.compilation.tap('RepackOutputPlugin', compilation => {
+      compilation.hooks.processAssets.tap({
+        name: 'RepackOutputPlugin',
+        stage: _webpack.default.Compilation.PROCESS_ASSETS_STAGE_ADDITIONS
+      }, () => {
+        var _entryGroup;
+
+        entryGroup = compilation.chunkGroups.find(group => group.isInitial());
+        entryChunk = (_entryGroup = entryGroup) === null || _entryGroup === void 0 ? void 0 : _entryGroup.chunks.find(chunk => chunk.name === entryChunkName);
+        const sharedChunks = new Set();
 
-      // Process shared chunks to add them either as local or remote chunk.
-      for (const sharedChunk of sharedChunks) {
-        const isUsedByLocalChunk = localChunks.some(localChunk => getAllInitialChunks(localChunk, statsChunkMap).includes(sharedChunk));
-        if (isUsedByLocalChunk || isLocalChunk(sharedChunk.name ?? sharedChunk.id?.toString())) {
-          localChunks.push(sharedChunk);
-        } else {
-          remoteChunks.push(sharedChunk);
+        for (const chunk of compilation.chunks) {
+          var _chunk$id;
+
+          // Do not process shared chunks right now.
+          if (sharedChunks.has(chunk)) {
+            continue;
+          }
+
+          [...chunk.getAllInitialChunks()].filter(sharedChunk => sharedChunk !== chunk).forEach(sharedChunk => {
+            sharedChunks.add(sharedChunk);
+          }); // Entry chunk
+
+          if (entryChunk && entryChunk === chunk) {
+            localChunks.push(chunk);
+          } else if (isLocalChunk(chunk.name ?? ((_chunk$id = chunk.id) === null || _chunk$id === void 0 ? void 0 : _chunk$id.toString()))) {
+            localChunks.push(chunk);
+          } else {
+            remoteChunks.push(chunk);
+          }
+        } // Process shared chunks to add them either as local or remote chunk.
+
+
+        for (const sharedChunk of sharedChunks) {
+          var _sharedChunk$id;
+
+          const isUsedByLocalChunk = localChunks.some(localChunk => {
+            return [...localChunk.getAllInitialChunks()].includes(sharedChunk);
+          });
+
+          if (isUsedByLocalChunk || isLocalChunk(sharedChunk.name ?? ((_sharedChunk$id = sharedChunk.id) === null || _sharedChunk$id === void 0 ? void 0 : _sharedChunk$id.toString()))) {
+            localChunks.push(sharedChunk);
+          } else {
+            remoteChunks.push(sharedChunk);
+          }
         }
+
+        if (!entryChunk) {
+          throw new Error('Cannot infer entry chunk - this should have not happened.');
+        } // Collect auxiliary assets (only remote-assets for now)
+
+
+        Object.keys(compilation.assets).filter(filename => /^remote-assets/.test(filename)).forEach(asset => auxiliaryAssets.add(asset));
+      });
+    });
+    compiler.hooks.afterEmit.tapPromise('RepackOutputPlugin', async compilation => {
+      var _localAssetsCopyProce2, _auxiliaryAssetsCopyP;
+
+      const outputPath = compilation.outputOptions.path;
+
+      if (!outputPath) {
+        throw new Error('Cannot infer output path from compilation');
       }
-      const assets = compilationStats.assets;
-      // Collect auxiliary assets (only remote-assets for now)
-      assets.filter(asset => /^remote-assets/.test(asset.name)).forEach(asset => auxiliaryAssets.add(asset.name));
+
       let localAssetsCopyProcessor;
       let {
         bundleFilename,
         sourceMapFilename,
         assetsPath
       } = this.config.output;
+
       if (bundleFilename) {
         if (!_path.default.isAbsolute(bundleFilename)) {
           bundleFilename = _path.default.join(this.config.context, bundleFilename);
         }
+
         const bundlePath = _path.default.dirname(bundleFilename);
+
         if (!sourceMapFilename) {
           sourceMapFilename = `${bundleFilename}.map`;
         }
+
         if (!_path.default.isAbsolute(sourceMapFilename)) {
           sourceMapFilename = _path.default.join(this.config.context, sourceMapFilename);
         }
+
         if (!assetsPath) {
           assetsPath = bundlePath;
         }
+
         logger.debug('Detected output paths:', {
           bundleFilename,
           bundlePath,
@@ -178,6 +182,7 @@ class OutputPlugin {
         });
         localAssetsCopyProcessor = new _AssetsCopyProcessor.AssetsCopyProcessor({
           platform: this.config.platform,
+          compilation,
           outputPath,
           bundleOutput: bundleFilename,
           bundleOutputDir: bundlePath,
@@ -186,24 +191,34 @@ class OutputPlugin {
           logger
         });
       }
+
       const remoteAssetsCopyProcessors = {};
+
       for (const chunk of localChunks) {
+        var _localAssetsCopyProce;
+
         // Process entry chunk
-        localAssetsCopyProcessor?.enqueueChunk(chunk, {
-          isEntry: entryChunk === chunk,
-          sourceMapFile: getRelatedSourceMap(chunk)
+        (_localAssetsCopyProce = localAssetsCopyProcessor) === null || _localAssetsCopyProce === void 0 ? void 0 : _localAssetsCopyProce.enqueueChunk(chunk, {
+          isEntry: entryChunk === chunk
         });
       }
+
       for (const chunk of remoteChunks) {
-        const spec = extraAssets.find(spec => _webpack.default.ModuleFilenameHelpers.matchObject({
-          test: spec.test,
-          include: spec.include,
-          exclude: spec.exclude
-        }, chunk.name || chunk.id?.toString()));
-        if (spec?.type === 'remote') {
+        const spec = extraAssets.find(spec => {
+          var _chunk$id2;
+
+          return _webpack.default.ModuleFilenameHelpers.matchObject({
+            test: spec.test,
+            include: spec.include,
+            exclude: spec.exclude
+          }, chunk.name || ((_chunk$id2 = chunk.id) === null || _chunk$id2 === void 0 ? void 0 : _chunk$id2.toString()));
+        });
+
+        if ((spec === null || spec === void 0 ? void 0 : spec.type) === 'remote') {
           if (!remoteAssetsCopyProcessors[spec.outputPath]) {
             remoteAssetsCopyProcessors[spec.outputPath] = new _AssetsCopyProcessor.AssetsCopyProcessor({
               platform: this.config.platform,
+              compilation,
               outputPath,
               bundleOutput: '',
               bundleOutputDir: spec.outputPath,
@@ -212,16 +227,18 @@ class OutputPlugin {
               logger
             });
           }
+
           remoteAssetsCopyProcessors[spec.outputPath].enqueueChunk(chunk, {
-            isEntry: false,
-            sourceMapFile: getRelatedSourceMap(chunk)
+            isEntry: false
           });
         }
       }
+
       let auxiliaryAssetsCopyProcessor;
       const {
         auxiliaryAssetsPath
       } = this.config.output;
+
       if (auxiliaryAssetsPath) {
         auxiliaryAssetsCopyProcessor = new _AuxiliaryAssetsCopyProcessor.AuxiliaryAssetsCopyProcessor({
           platform: this.config.platform,
@@ -229,13 +246,17 @@ class OutputPlugin {
           assetsDest: auxiliaryAssetsPath,
           logger
         });
+
         for (const asset of auxiliaryAssets) {
           auxiliaryAssetsCopyProcessor.enqueueAsset(asset);
         }
       }
-      await Promise.all([...(localAssetsCopyProcessor?.execute() ?? []), ...Object.values(remoteAssetsCopyProcessors).reduce((acc, processor) => acc.concat(...processor.execute()), []), ...(auxiliaryAssetsCopyProcessor?.execute() ?? [])]);
+
+      await Promise.all([...(((_localAssetsCopyProce2 = localAssetsCopyProcessor) === null || _localAssetsCopyProce2 === void 0 ? void 0 : _localAssetsCopyProce2.execute()) ?? []), ...Object.values(remoteAssetsCopyProcessors).reduce((acc, processor) => acc.concat(...processor.execute()), []), ...(((_auxiliaryAssetsCopyP = auxiliaryAssetsCopyProcessor) === null || _auxiliaryAssetsCopyP === void 0 ? void 0 : _auxiliaryAssetsCopyP.execute()) ?? [])]);
     });
   }
+
 }
+
 exports.OutputPlugin = OutputPlugin;
 //# sourceMappingURL=OutputPlugin.js.map
\ No newline at end of file

```

**Comment by jbroma** [comment]
this should be fixed by #646 (you can try using that as a patch and see if it works for your case)

**Comment by madaky** [comment]
@jbroma : Any plan to release this, or can you please suggest a alternative solution for this.
